packages/app/src/systems/Swap/machines/swapMachine.ts

```diff
@@ -0,0 +1,186 @@
+    },
+    fetchingPreview: {},
+  },
+  on: {
+    SET_DIRECTION: {
+      actions: ['setDirection'],
+    },
+    INVERT_COINS: {
+      actions: ['invertDirection'],
+      target: 'fetchingBalances',
+    },
+    SELECT_COIN: {
+      actions: ['selectCoin'],
+      target: 'fetchingBalances',
+    },
+  },
+});
+
+export const swapMachine = machine.withConfig({
+  services: {
+    fetchBalances: async (ctx) => {
+      const { coinFrom, coinTo, wallet } = ctx;
+      const fromID = coinFrom?.assetId;
+      const toID = coinTo?.assetId;
+      const fromBalance = coinFrom && (await wallet?.getBalance(fromID));
+      const toBalance = coinTo && (await wallet?.getBalance(toID));
+
+      return {
+        fromBalance,
+        toBalance,
+      };
+    },
+    fetchTxCost: async (ctx) => {
+      if (!ctx.contract) {
+        throw new Error('Contract not found');
+      }
+      const networkFee = queryNetworkFeeOnSwap(ctx.contract, ctx.direction);
+      const txCost = getTransactionCost(networkFee);
+      return txCost || emptyTransactionCost();
+    },
+    fetchPoolRatio: async (ctx) => {
+      if (!ctx.contract) {
+        throw new Error('Contract not found');
+      }
+      const info = await ctx.contract?.dryRun.get_pool_info();
+      const ratio = await getPoolRatio(info);
+      return { info, ratio };
+    },
+  },
+  actions: {
+    setBalances: assign((ctx, ev) => ({
+      ...ctx,
+      coinFromBalance: ev.data.fromBalance,
+      coinToBalance: ev.data.toBalance,
+    })),
+    setTxCost: assign({
+      txCost: (_, ev) => ev.data as TransactionCost,
+    }),
+    setPoolInfo: assign({
+      poolInfo: (_, ev) => ev.data.info,
+    }),
+    setPoolRatio: assign({
+      poolRatio: (_, ev) => ev.data.ratio,
+    }),
+    selectCoin: assign((ctx, ev) => ({
+      ...ctx,
+      ...(ev.data.direction === FROM_TO && {
+        coinFrom: ev.data.coin,
+      }),
+      ...(ev.data.direction === TO_FROM && {
+        coinTo: ev.data.coin,
+      }),
+    })),
+    invertDirection: assign((ctx) => {
+      const isFrom = ctx.direction === FROM_TO;
+      return {
+        direction: isFrom ? TO_FROM : FROM_TO,
+        coinFrom: ctx.coinTo,
+        coinTo: ctx.coinFrom,
+        fromAmount: ctx.toAmount,
+        toAmount: ctx.fromAmount,
+      };
+    }),
+    setDirection: assign({
+      direction: (_, ev) => ev.data,
+    }),
+    setActiveValue: assign((ctx, ev) => {
+      const isFrom = ctx.direction === FROM_TO;
+      const next = {
+        value: ev.data,
+        raw: parseValueBigInt(ev.data),
+      };
+      return {
+        ...ctx,
+        ...(isFrom ? { fromAmount: next } : { toAmount: next }),
+      };
+    }),
+  },
+});
+
+export type SwapMachine = typeof swapMachine;
+export type SwapMachineService = InterpreterFrom<SwapMachine>;
+export type SwapMachineState = StateFrom<SwapMachine>;
+
+export function isLoadingState(state: SwapMachineState) {
+  return (
+    state.matches('fetchingBalances') ||
```

Comment:
It's possible to use enums instead of the strings here and replace them on the machine with the enum also like:
```ts
{
  [SwapStates.FetchingTxCost]: {...}
}
```

Reply:
That's a session talking about this on XState docs
https://xstate.js.org/docs/guides/typescript.html#don-t-use-enums